include/swift/ABI/SuppressibleProtocols.h

```diff
@@ -0,0 +1,224 @@
+//===--- SuppressibleProtocols.h - Suppressible protocol meta ---*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This header declares various types for suppressible protocols, such as
+// Copyable.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_ABI_SUPPRESSIBLEPROTOCOLS_H
+#define SWIFT_ABI_SUPPRESSIBLEPROTOCOLS_H
+
+#include <cstdint>
+
+namespace swift {
+
+/// Describes a "suppressible" protocol, such as Copyable, which is assumed to
+/// hold for all types unless explicitly suppressed.
+enum class SuppressibleProtocolKind : uint8_t {
+#define SUPPRESSIBLE_PROTOCOL(Name, Bit, MangleChar) Name = Bit,
+#include "swift/ABI/SuppressibleProtocols.def"
+};
+
+/// A set of suppressible protocols, whose bits correspond to the cases of
+/// SuppressibleProtocolKind.
+class SuppressibleProtocolSet {
+  using StorageType = uint16_t;
+  
+  /// The stored bits.
+  StorageType bits;
+
+  /// Retrieve the mask for this bit.
+  static StorageType getMask(SuppressibleProtocolKind kind) {
+    return 1 << static_cast<uint8_t>(kind);
+  }
+  
+public:
+  explicit constexpr SuppressibleProtocolSet(StorageType bits) : bits(bits) {}
+  constexpr SuppressibleProtocolSet() : bits(0) {}
+
+  /// Retrieve the raw bits that describe this set.
+  StorageType rawBits() const { return bits; }
+
+  /// Whether the set contains no protocols.
+  bool empty() const { return bits == 0; }
+
+  /// Check whether the set contains the specific suppressible protocol.
+  constexpr bool contains(SuppressibleProtocolKind kind) const {
+    return bits & getMask(kind);
+  }
+
+  /// Insert the suppressible protocol into the set.
+  void insert(SuppressibleProtocolKind kind) {
+    bits = bits | getMask(kind);
+  }
+
+  /// Remove the given suppressible protocol from the set.
+  void remove(SuppressibleProtocolKind kind) {
+    uint16_t mask = getMask(kind);
+    bits = bits & ~mask;
+  }
+
+  /// Clear out all of the protocols from the set.
+  void clear() { bits = 0; }
+
+#define SUPPRESSIBLE_PROTOCOL(Name, Bit, MangleChar) \
+  constexpr bool contains##Name() const {            \
+    return contains(SuppressibleProtocolKind::Name); \
+  }
+#include "swift/ABI/SuppressibleProtocols.def"
```

Comment:
A static assert that the value for `Bit` fits into `StorageType` might be nice. Unlikely we'd ever hit it, but would save some pain in the off chance we did.

Reply:
Sure. I'll bury it in a C++ file somewhere